rbd/rbd.go

```diff
@@ -1286,3 +1259,103 @@ func TrashRestore(ioctx *rados.IOContext, id, name string) error {
 
 	return GetError(C.rbd_trash_restore(C.rados_ioctx_t(ioctx.Pointer()), c_id, c_name))
 }
+
+// OpenImage will open an existing rbd image by name and snapshot name,
+// returning a new opened image. Pass the NoSnapshot sentinel value as the
+// snapName to explicitly indicate that no snapshot name is being provided.
+//
+// Implements:
+//  int rbd_open(rados_ioctx_t io, const char *name,
+//               rbd_image_t *image, const char *snap_name);
+func OpenImage(ioctx *rados.IOContext, name, snapName string) (*Image, error) {
```

Comment:
There is also https://godoc.org/google.golang.org/cloud/internal/optional#String , but i don't think it would add a lot value compared to a pointer.

Reply:
I have to disagree on this one. After using it with the sentinel value `NoSnapshot` instead of empty string it flows pretty nice and still matches well with the underlying apis. Note that if there's a lot of confusion in the future doing it with a sentinel value doesn't preclude adding some sort of higher level interface on top.
I'll also note that I checked the ceph sources and empty string is treated the same as null in the C++ code, so even though we're being extra careful and sticking with the documented approach to pass "NULL for no snapshot" to convert to a null its not totally needed.

Reply:
Then we have to agree to disagree. :wink: For me "abusing" an empty string is kind of dangerous in-band signalling. I looked for an option or maybe type, that exists in other languages, but it seems the "Go idiomatic" way to do it is pointers, which I also don't find exciting, but since they are reference counted it's a lot less dangerous than in C.

Just because the current C API accepts undocumented parameters, doesn't mean that it could be changed anytime, so for me that is a dangerous argument. Also, I have a gut feeling, that using a constant from the package namespace might have other unwanted effects. Would that needed to be referred to as rbd.NoSnapshot when called from outside? Is this a standard pattern in Go APIs?

And we anyway have to document the sentinel, right? So we also don't save effort there, compared to documenting the meaning of a `nil` or an empty string.

On the other hand, I see that this is clearly in the area of academic discussions, so I would not veto even if i could. But my experience says that API design is really not trivial, because you want to change it as rarely as possible, so it's worth it to think three times about it, and to check what others do.